service_profiles.md

```diff
@@ -0,0 +1,111 @@
+for Linkerd to retry.  The types of characteristics that could potentially be
+described by a service profile includes:
+
+* success classification
+* retryability/idempotence
+* retry budgets
+* rate limits
+* ACLs
+* latency objectives/timeouts
+* success rate objectives
+* etc.
+
+Each service can be associated with at most one profile.  While we may one day
+wish to explore extending or overriding profiles, this is currently out of
+scope.  It is important to note that even though it is a service that is
+associated with a profile, this can influence the behavior of callers of that
+service.  In many cases the service and its caller may exist in different
+namespaces and/or have different owners.
+
+Service profiles are designed to be built in an incremental and extensible way.
+The initial implementation will contain very few characteristics but will be
+expanded over time.
+
+## Service Profile Definition
+
+While some characteristics such as retry budgets apply to the service as a
+whole, many characteristics vary depending on the request.  For example,
+idempotence may vary depending on the request's HTTP method or timeouts might
+vary depending on the request's HTTP path.
+
+To accommodate this, a service profile contains some top level characteristics
+that apply to the whole service and an ordered list of request policy objects
+which define the characteristics which apply to certain classes of requests.
+
+```
+message ServiceProfile {
+  repeated RequestPolicy request_policies = 1;
+  RetryBudget retry_budget = 2;
+  Duration default_timeout = 3;
+  // More service level characteristics to be added in the future
+}
+```
+
+Each request policy object defines a request match condition that describes
+which requests fall under this policy.  If a request matches more than one
+request policy object's condition, only the first match is considered.  The
+request policy object also defines the characteristics that apply to its
+requests.  For example, it defines a response match condition that describes
+which responses should be considered successful and another response match
+condition that describes which responses should be considered retryable if they
+are not successful.
+
+```
+message RequestPolicy {
+  RequestMatch condition = 1;
+  ResponseMatch success_condition = 2;
+  ResponseMatch retry_condition = 3;
+  Duration timeout = 4;
+  // More request policy characteristics to be added in the future
+}
+```
+
+For example, consider the HTTP REST service profile:
+
+```
+ServiceProfile {
+  request_policies: [
+    RequestPolicy {
+      // idempotent requests
+      condition: Any(Method(GET), Method(PUT), Method(DELETE), Method(HEAD), Method(TRACE))
+      success_condition: Status < 500
+      retry_condition: Always
+    },
+    RequestPolicy {
+      // fallback, (non-idempotent)
+      condition: Always
+      success_condition: Status < 500
+      retry_condition: Never
```

Comment:
I think I can see why having a `Never` makes sense as something *higher* priority, (say, PUTs with a 500 are dangerous in my system, but all other idempotent methods with any 5xx are safe), but is it necessary to have a fallback? I'd assume that if the list of policies has been exhausted and nothing matched, the proxy can just assume that it cannot retry.

Reply:
Yeah, depending on what the default service profile is, this fallback might be redundant.  What is listed there in the fallback is probably a good default (for HTTP).

Reply:
I think I forgot that this was also supposed to be used to give better metrics, so even though the fallback would not be to retry, it could still configure how to classify a success.

Reply:
Yeah, though the success function defined there is probably the default success function for HTTP anyway.